src/bio/bio_buffer.c

```diff
@@ -176,23 +176,24 @@ bio_iod_sgl(struct bio_desc *biod, unsigned int idx)
 struct bio_desc *
 bio_iod_alloc(struct bio_io_context *ctxt, unsigned int sgl_cnt, bool update)
 {
-	struct bio_desc *biod;
+	void		*alloc_buf;
+	daos_size_t	 biod_size, bsgls_size;
+	struct bio_desc	*biod;
+
 	D_ASSERT(ctxt != NULL && ctxt->bic_umem != NULL);
 	D_ASSERT(sgl_cnt != 0);
 
-	D_ALLOC_PTR(biod);
-	if (biod == NULL)
+	biod_size = roundup(sizeof(*biod), 8);
+	bsgls_size = sizeof(*biod->bd_sgls) * sgl_cnt;
+	D_ALLOC(alloc_buf, biod_size + bsgls_size);
+	if (alloc_buf == NULL)
 		return NULL;
 
+	biod = alloc_buf;
 	biod->bd_ctxt = ctxt;
 	biod->bd_update = update;
 	biod->bd_sgl_cnt = sgl_cnt;
-
-	D_ALLOC_ARRAY(biod->bd_sgls, sgl_cnt);
-	if (biod->bd_sgls == NULL) {
-		D_FREE(biod);
-		return NULL;
-	}
+	biod->bd_sgls = alloc_buf + biod_size;
```

Comment:
This looks complicated to me, to make bd_sgls as inline array for reducing one allocation, I think the common way is to move the bd_sgls to the end of bio_desc, and define it as bd_sgls[0], then we here just need a single call "D_ALLOC(biod, sizeof(*biod) + sizeof(struct bio_sgls) * sgl_cnt)"

Reply:
Ah, yes. it is because I added another array (bholes_sizes) in my first version so write as that way, but forget to change after I remove that extra array.
let me update it.